### ISSUE TITLE
Enabling multi-gpu training

### ISSUE THREAD

**Issue by Nav33nn** [allenai/allennlp#1882]
**Is your feature request related to a problem? Please describe.**
We have been working on the allennlp bi-DAF model for machine comprehension and it has come to our noticed that the multi-gpu code is a work in progress. 

**Describe alternatives you've considered**
We have found a work around from the issue #1439 but the problem with that is everytime the library is updated, we will have to make those changes in our code and rebuilld the whole thing.

**Describe the solution you'd like**
Since we already have our code using allennlp library, we would like an attempt to fix this problem hoping that the we will not have to touch the multi gpu code again when there is a new release. We saw from your latest issues that this is being worked on and would like to know if we could expedite this feature by contributing to it. Do let us know if there are any specific issues we could pick up with regards to this. @matt-gardner 

Thanks,

**Comment by matt-gardner** [comment]
I'd say generally, yes, contributions are very welcome.  If your problem is just what was described in #1439, there was some work trying to fix in it #1724, but that PR has stalled (not sure if @vidurj is still working on it or not; probably not, because of the inactivity there).  There was some discussion on that issue about how to solve the problem in a general way (see [this comment](https://github.com/allenai/allennlp/pull/1724#discussion_r215810380)) - basically we need to pull in the `scatter_kwargs` function from `from torch.nn.parallel.scatter_gather` with our own implementation that does the right thing on our data types.

**Comment by brendan-ai2** [comment]
Hi @Nav33nn, there is now rudimentary support for multi-GPU training as of https://github.com/allenai/allennlp/pull/1944. Could you please give that a try and let me know if you have any issues? Thanks!

**Comment by Nav33nn** [comment]
@brendan-ai2 that's great news. Will get back in a weeks time. Thanks!

**Comment by vidurj** [comment]
@sai-prasanna I would expect using multiple GPUs to process the same total batch size to be slower than using a single gpu if the batch fits on 1 gpu.

**Comment by Nav33nn** [comment]
@brendan-ai2: Thanks for the fix! Validated multi-GPU on bi-daf pulled from #1944. Works fine for me.

**Comment by sai-prasanna** [comment]
> @sai-prasanna My hunch is that using multiple GPUs to process the same total batch size will be slower than using a single gpu if the batch fits on 1 gpu.

But even after doubling the batch size it doesn't improve the speed.

**Comment by matt-gardner** [comment]
@sai-prasanna, what model are you using?  We are still working on making multi-gpu support work well, so it's not surprising to me that you're not seeing gains here.  In particular, we have a [multi-process data iterator](https://github.com/allenai/allennlp/blob/master/allennlp/data/iterators/multiprocess_iterator.py) that I'd expect you'd need to use to see gains when you're CPU-bound and not GPU-bound (which looks to be the case, if a batch size of 16 takes the same amount of time as a batch size of 32).

**Comment by sai-prasanna** [comment]
@matt-gardner  I was using a sequence 2 sequence model, importing the encoder and decoder from fairseq library. 
Know you all are working on it, I  was just doing some preliminary experimenting with it (got 2 more GPUs :)).

**Comment by matt-gardner** [comment]
Ok, thanks for the data point, hearing other people's experiences trying to use it is helpful.  Did you try the multi-process data iterator?

**Comment by brendan-ai2** [comment]
Thanks for the feedback, @Nav33nn!

@sai-prasanna, as Matt mentioned you should be able to see gains now if you're GPU bound. The GIL should not be an issue in this case. (See https://pytorch.org/docs/stable/notes/cuda.html#use-nn-dataparallel-instead-of-multiprocessing) As a sanity check, is 32 the largest batch you can fit on a single GPU?

**Comment by sai-prasanna** [comment]
@brendan-ai2  - I can fit 50 instances per batch in 1 GPU. So I tried rerunning the experiment, with and without multiprocess iterator. Had to set ulimit -n 9000 to avoid some issues.

Here are some snapshots of the training progress.

Without Multi Process Iterator - 1 GPU -  Batch Size 50

> loss: 4.9577 ||:   1%|          | 295/55226 [04:44<1:01:48, 14.81it/s]


Without Multi Process Iterator - 2 GPU - - Batch Size 100

> loss: 3.6766 ||:   2%|2         | 622/27613 [07:13<2:11:33,  3.42it/s]

Without Multi Process Iterator - 2 GPU - - Batch Size 100

Here it is wierd, the total batches is now showing '86290' eventhough it should show 27k for the same dataset with batch size 100.

> loss: 2.5587 ||:   2%|2         | 2046/86290 [12:48<9:42:45,  2.41it/s]

**Comment by schmmd** [comment]
The OP's issue is resolve.  @sai-prasanna are you still working through this problem?

**Comment by sai-prasanna** [comment]
My friend @scarecrow1123  is experimenting on using distributed data parallel for training. We will hopefully get it ready soon.

**Comment by matt-gardner** [comment]
I'm pretty sure #1890 is tracking what is left of this issue, so I'm going to close this one.
